crates/signature-mpc/src/twopc_mpc_protocols/decrypt.rs

```diff
@@ -0,0 +1,198 @@
+        PublicNonceEncryptedPartialSignatureAndProof<ProtocolContext>,
+    >,
+    signature_threshold_decryption_round_parties: Vec<SignatureThresholdDecryptionParty>,
+    messages: Vec<Vec<u8>>,
+) -> Result<Vec<Vec<u8>>, FailedMessagesIndices> {
+    let mut failed_messages_indices = Vec::new();
+    let messages_signatures: Vec<Vec<u8>> = signature_threshold_decryption_round_parties
+        .into_iter()
+        .zip(
+            messages
+                .into_iter()
+                .zip(public_nonce_encrypted_partial_signature_and_proofs.into_iter())
+                .zip(decryption_shares.iter()),
+        )
+        .enumerate()
+        .map(
+            |(
+                index,
+                (
+                    signature_threshold_decryption_round_party,
+                    (
+                        (message, public_nonce_encrypted_partial_signature_and_proof),
+                        (partial_signature_decryption_shares, masked_nonce_decryption_shares),
+                    ),
+                ),
+            )| {
+                let result = signature_threshold_decryption_round_party.decrypt_signature(
+                    lagrange_coefficients.clone(),
+                    partial_signature_decryption_shares.clone(),
+                    masked_nonce_decryption_shares.clone(),
+                );
+
+                match result {
+                    Ok((nonce_x_coordinate, signature_s)) => {
+                        let signature_s_inner: k256::Scalar = signature_s.into();
+                        Signature::<k256::Secp256k1>::from_scalars(
+                            k256::Scalar::from(nonce_x_coordinate),
+                            signature_s_inner,
+                        )
+                        .unwrap()
+                        .to_vec()
+                    }
+                    Err(_) => {
+                        failed_messages_indices.push(index);
+                        Vec::new()
+                    }
+                }
+            },
+        )
+        .collect();
+
+    if !failed_messages_indices.is_empty() {
+        return Err(failed_messages_indices);
+    }
+    Ok(messages_signatures)
+}
+
+pub fn decrypt_signature_decentralized_party_sign(
+    messages: Vec<Vec<u8>>,
+    decryption_key_share_public_parameters: DecryptionPublicParameters,
+    decryption_shares: HashMap<PartyID, Vec<(PaillierModulusSizedNumber, PaillierModulusSizedNumber)>>,
+    public_nonce_encrypted_partial_signature_and_proofs: Vec<
+        PublicNonceEncryptedPartialSignatureAndProof<ProtocolContext>,
+    >,
+    signature_threshold_decryption_round_parties: Vec<SignatureThresholdDecryptionParty>,
+) -> Result<Vec<Vec<u8>>, DecryptionError> {
+    let (decrypters, decryption_shares) = take_threshold_decrypters(
+        decryption_shares,
+        public_nonce_encrypted_partial_signature_and_proofs.clone(),
+        decryption_key_share_public_parameters.clone(),
+    );
```

Comment:
Who said its going to be the same decrypters for all messages?

Reply:
Theoretically, once we have threshold decrypters, we should start decrypting the signature so that it will be the same decrypters..?